File: RushTodo.Api/Validators/ValidatorBase.cs
using RushTodo.Api.Helpers;
using RushTodo.Api.Models;

namespace RushTodo.Api.Validators;

public abstract class ValidatorBase
{
    public string Label { get; init; } = "This field";
    public bool IsEditable { get; init; } = true;
    public bool IsVisible { get; init; } = true;

    public virtual string? Validate() => null;

    public void Assert()
    {
        var validation = Validate();
        if (validation is not null) throw new ValidationException(validation);
    }
}


public abstract class ValidatorBase<TModel>(TModel model) : ValidatorBase
    where TModel : Model
{
    protected TModel Model { get; } = model;
}


public class ValueValidator<T> : ValidatorBase
{
    public Func<T>? GetValue { get; init; }
    public Func<T, string?>? Validation { get; init; }

    public override string? Validate()
    {
        if (!IsVisible || GetValue is null) return null;

        var value = GetValue();
        return Validate(value)?.Replace("[Label]", Label);
    }

    public string? Validate(T value)
        => IsVisible ? Validation?.Invoke(value)?.Replace("[Label]", Label) : null;
}


public class DateValidator : ValueValidator<DateOnly?>
{
    public DateOnly? Min { get; init; }
    public DateOnly? Max { get; init; }
}


public class TextValidator : ValueValidator<string?>
{
    public int? MaxLength { get; init; }
}

File: rush-todo-web/src/validators/ValidatorBase.ts
type ValueValidatorOptions<T> = {
  getValue: () => T
  isVisible?: boolean
  label: string
  validation?: (value: T) => string | undefined
}

abstract class ValidatorBase {
  readonly isVisible: boolean
  readonly label: string

  constructor(label: string, isVisible = true) {
    this.isVisible = isVisible
    this.label = label
  }

  abstract validate(): string | undefined
}

class ValueValidator<T> extends ValidatorBase {
  private readonly getValue: () => T
  private readonly validation?: (value: T) => string | undefined

  constructor(options: ValueValidatorOptions<T>) {
    super(options.label, options.isVisible)
    this.getValue = options.getValue
    this.validation = options.validation
  }

  override validate(): string | undefined {
    if (!this.isVisible) return undefined

    return this.validation?.(this.getValue())?.replace('[Label]', this.label)
  }
}

type TextValidatorOptions = ValueValidatorOptions<string | null> & {
  maxLength?: number
}

class TextValidator extends ValueValidator<string | null> {
  readonly maxLength?: number

  constructor(options: TextValidatorOptions) {
    super(options)
    this.maxLength = options.maxLength
  }
}

class DateValidator extends ValueValidator<string | null> {}

export { DateValidator, TextValidator, ValidatorBase, ValueValidator }
